components/builder-api/src/server/resources/channels.rs

```diff
@@ -31,354 +33,314 @@ impl Channels {
-    request.set_visibilities(all_visibilities());
+    //
+    // Route handlers - these functions should return HttpResponse
+    //
+    pub fn list_channels(req: &HttpRequest<AppState>) -> HttpResponse {
+        let origin = Path::<(String)>::extract(&req).unwrap().into_inner(); //Unwrap apparently Ok
 
-    let package = match route_message::<OriginPackageGet, OriginPackage>(req, &request) {
-        Ok(p) => p,
-        Err(e) => {
-            error!("Error retrieving package: {:?}", &e);
-            return Ok(render_net_error(&e));
+        let mut request = OriginChannelListRequest::new();
+        request.set_include_sandbox_channels(false);
+
+        match helpers::get_origin(req, &origin) {
+            Ok(orgn) => request.set_origin_id(orgn.get_id()),
+            Err(err) => return err.into(),
         }
-    };
-
-    let mut promote = OriginPackagePromote::new();
-    promote.set_channel_id(origin_channel.get_id());
-    promote.set_package_id(package.get_id());
-    promote.set_ident(ident.clone());
-
-    match route_message::<OriginPackagePromote, NetOk>(req, &promote) {
-        Ok(_) => match route_message::<OriginPackagePromote, NetOk>(req, &promote) {
-            Ok(_) => match audit_package_rank_change(
-                req,
-                package.get_id(),
-                origin_channel.get_id(),
-                PackageChannelOperation::Promote,
-                origin_id,
-            ) {
-                Ok(_) => Ok(Response::with(status::Ok)),
-                Err(err) => return Ok(render_net_error(&err)),
-            },
-            Err(err) => return Ok(render_net_error(&err)),
-        },
-        Err(err) => Ok(render_net_error(&err)),
-    }
-}
 
-fn demote_package(req: &mut Request) -> IronResult<Response> {
-    let mut ident = OriginPackageIdent::new();
-    match get_param(req, "origin") {
-        Some(origin) => ident.set_origin(origin),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    match get_param(req, "pkg") {
-        Some(pkg) => ident.set_name(pkg),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    match get_param(req, "version") {
-        Some(version) => ident.set_version(version),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    match get_param(req, "release") {
-        Some(release) => ident.set_release(release),
-        None => return Ok(Response::with(status::BadRequest)),
-    }
-    let channel = match get_param(req, "channel") {
-        Some(channel) => channel,
-        None => return Ok(Response::with(status::BadRequest)),
-    };
-
-    // you can't demote from "unstable"
-    if channel == "unstable" {
-        return Ok(Response::with(status::Forbidden));
+        // Pass ?sandbox=true to this endpoint to include sandbox channels in the list. They are not
+        // there by default.
+        let sandbox = Query::<(String)>::extract(&req).unwrap().into_inner();
```

Comment:
You can put this query param handling in the function prototype, and have it automatically parse into a struct... see the code sample from one of the other modules (eg pkgs.rs).. lot cleaner